test/req_test.exs

```diff
@@ -95,4 +95,12 @@ defmodule ReqTest do
     resp = Req.get!(origin_url, into: :self)
     assert Req.put!(echo_url, body: resp.body).body == "foobarbaz"
   end
+
+  test "empty params", c do
+    Bypass.expect(c.bypass, "GET", "/", fn conn ->
+      Plug.Conn.send_resp(conn, 200, "")
+    end)
+
+    assert %Req.Response{status: 200} = Req.get!(c.url, params: nil)
+  end
```

Comment:
Hi @wojtekmach I think it's a valid concern. IMO, it depends on how you want to communicate "emptiness". I usually read something like `[]` or `%{}` as empty and `nil` as uninitialized / invalid. 

For instance, it seems that Plug handles `%{}`, `[]` and `nil` as a no-op:

```elixir
Plug.Conn.fetch_query_params(Plug.Adapters.Test.Conn.conn(%Plug.Conn{}, :get, "/", nil))
# => %Plug.Conn{query_params: %{}, ...}
Plug.Conn.fetch_query_params(Plug.Adapters.Test.Conn.conn(%Plug.Conn{}, :get, "/", %{}))
# => %Plug.Conn{query_params: %{}, ...} 
Plug.Conn.fetch_query_params(Plug.Adapters.Test.Conn.conn(%Plug.Conn{}, :get, "/", []))
# => %Plug.Conn{query_params: %{}, ...}  
```

However, if you want to stay closer to `Map.merge/2` semantics, it makes total sense to accept `nil`. 

FYI: The code below also raises, so merging params to nil is already invalid:

```elixir
Req.get(Req.new(url: c.url, params: [a: 1]), params: nil)
```

I just pushed a new commit that handles the proper merging of `nil` params and coalesces while encoding the query params instead no-oping when you pass `nil`: https://github.com/wojtekmach/req/pull/423/commits/1338d4dfbb5c1a4fc22ecf767ef9f96541b9f4cb.

Reply:
Thank you. It's always a good reminder what Plug and similar libraries do. I'd like to take some time thinking this through. FWIW we could raise a good error message for `nil` telling users to change their code, for example, nudging users to make a change like this:

```diff
- Req.get(Req.new(url: "/", params: opts[:query]))
+ Req.get(Req.new(url: "/", params: opts[:query] || []))
```

Reply:
My personal opinion is that this is such a common (expected) use case that I'd rather avoid `raise`.
IMHO coalescing the value in the lib code is a more elegant solution than branching on the client.

Let me know if there's something I can do to help because the current behavior is kinda annoying and will catch a lot of people off guard 😅 💪